File: PhotoMap/PhotosViewController/PhotosViewController.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 The secondary view controller used for browing the photos.
 */

#import "PhotosViewController.h"
#import "ModelController.h"
#import "DataViewController.h"

@interface PhotosViewController ()

@property (strong, nonatomic) ModelController *modelController;
@property (strong, nonatomic) UIPageViewController *pageViewController;

@end


#pragma mark -

@implementation PhotosViewController

- (void)updateNavBarTitle {
    
    if (self.modelController.pageData.count > 1) {
        self.title = [NSString stringWithFormat:@"Photos (%zd of %zd)", self.modelController.currentPageIndex + 1, (long)self.modelController.pageData.count];
    }
    else {
        DataViewController *viewController = [self.modelController.pageData objectAtIndex:self.modelController.currentPageIndex];
        self.title = viewController.title;
    }
}

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
	// Do any additional setup after loading the view, typically from a nib.
    // Configure the page view controller and add it as a child view controller.
    _pageViewController =
        [[UIPageViewController alloc] initWithTransitionStyle:UIPageViewControllerTransitionStylePageCurl
                                        navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal
                                                      options:nil];
    self.pageViewController.delegate = self;
    
    self.modelController.pageData = self.photosToShow;
    
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Storyboard" bundle:nil];
    DataViewController *startingViewController = [self.modelController viewControllerAtIndex:0 storyboard:storyboard];
    NSArray *viewControllers = @[startingViewController];
    [self.pageViewController setViewControllers:viewControllers
                                      direction:UIPageViewControllerNavigationDirectionForward
                                       animated:NO
                                     completion:nil];
    
    [self updateNavBarTitle];

    self.pageViewController.dataSource = self.modelController;
    
    [self addChildViewController:self.pageViewController];
    [self.view addSubview:self.pageViewController.view];
    [self.pageViewController didMoveToParentViewController:self];
    
    _pageAnimationFinished = YES;
}

- (ModelController *)modelController {
    
    // Return the model controller object, creating it if necessary.
    // In more complex implementations, the model controller may be passed to the view controller.
    if (_modelController == nil) {
        _modelController = [[ModelController alloc] init];
    }
    return _modelController;
}


#pragma mark - UIPageViewControllerDelegate

- (void)pageViewController:(UIPageViewController *)pageViewController didFinishAnimating:(BOOL)finished previousViewControllers:(NSArray *)previousViewControllers transitionCompleted:(BOOL)completed {
    
    // update the nav bar title showing which index we are displaying
    [self updateNavBarTitle];
    
    _pageAnimationFinished = YES;
}

- (UIPageViewControllerSpineLocation)pageViewController:(UIPageViewController *)pageViewController spineLocationForInterfaceOrientation:(UIInterfaceOrientation)orientation {
    
    // Set the spine position to "min" and the page view controller's view controllers array to contain just one view controller.
    // Setting the spine position to 'UIPageViewControllerSpineLocationMid' in landscape orientation sets the doubleSided property to YES, so set it to NO here.
    UIViewController *currentViewController = self.pageViewController.viewControllers[0];
    NSArray *viewControllers = @[currentViewController];
    [self.pageViewController setViewControllers:viewControllers
                                      direction:UIPageViewControllerNavigationDirectionForward
                                       animated:YES
                                     completion:nil];
    
    self.pageViewController.doubleSided = NO;
    return UIPageViewControllerSpineLocationMin;
}

- (void)pageViewController:(UIPageViewController *)pageViewController willTransitionToViewControllers:(NSArray *)pendingViewControllers
{
    _pageAnimationFinished = NO;
}

@end

File: PhotoMap/PhotosViewController/PhotosViewController.swift
//
//  PhotosViewController.swift
//  PhotoMap
//
//  Translated by OOPer in cooperation with shlab.jp, on 2014/11/23.
//
//
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information

 Abstract:
 The secondary view controller used for browing the photos.
 */

import UIKit

@objc(PhotosViewController)
class PhotosViewController: UIViewController, UIPageViewControllerDelegate {
    
    var photosToShow: [PhotoAnnotation] = []
    var pageAnimationFinished: Bool = false
    
    private lazy var modelController: ModelController = ModelController()
    private var pageViewController: UIPageViewController?
    
    
    //#MARK: -
    
    private func updateNavBarTitle() {
        
        if self.modelController.pageData.count > 1 {
            self.title = "Photos (\(self.modelController.currentPageIndex + 1) of \(self.modelController.pageData.count))"
        } else {
            let viewController = self.modelController.pageData[self.modelController.currentPageIndex] //Original sample code's bug?
            self.title = viewController.title
        }
    }
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        // Do any additional setup after loading the view, typically from a nib.
        // Configure the page view controller and add it as a child view controller.
        pageViewController =
            UIPageViewController(transitionStyle: .pageCurl,
                navigationOrientation: .horizontal,
                options: nil)
        self.pageViewController!.delegate = self
        
        self.modelController.pageData = self.photosToShow
        
        let storyboard = UIStoryboard(name: "Storyboard", bundle: nil)
        let startingViewController = self.modelController.viewControllerAtIndex(0, storyboard: storyboard)!
        let viewControllers = [startingViewController]
        self.pageViewController!.setViewControllers(viewControllers,
            direction: .forward,
            animated: false,
            completion: nil)
        
        self.updateNavBarTitle()
        
        self.pageViewController!.dataSource = self.modelController
        
        self.addChildViewController(self.pageViewController!)
        self.view.addSubview(self.pageViewController!.view)
        self.pageViewController!.didMove(toParentViewController: self)
        
        // add the page view controller's gesture recognizers to the book view controller's view
        // so that the gestures are started more easily
        self.view.gestureRecognizers = self.pageViewController!.gestureRecognizers
        
        pageAnimationFinished = true
    }
    
    // Return the model controller object, creating it if necessary.
    // In more complex implementations, the model controller may be passed to the view controller.
    
    
    //#MARK: - UIPageViewControllerDelegate
    
    func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
        
        // update the nav bar title showing which index we are displaying
        self.updateNavBarTitle()
        
        pageAnimationFinished = true
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, spineLocationFor orientation: UIInterfaceOrientation) -> UIPageViewControllerSpineLocation {
        
        // Set the spine position to "min" and the page view controller's view controllers array to contain just one view controller. Setting the spine position to 'UIPageViewControllerSpineLocationMid' in landscape orientation sets the doubleSided property to YES, so set it to NO here.
        let currentViewController = self.pageViewController!.viewControllers![0]
        let viewControllers = [currentViewController]
        self.pageViewController!.setViewControllers(viewControllers,
            direction: .forward,
            animated: true,
            completion: nil)
        
        self.pageViewController!.isDoubleSided = false
        return .min
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, willTransitionTo pendingViewControllers: [UIViewController]) {
        pageAnimationFinished = false
    }
    
}
